### DISCUSSION TITLE
9.x: Logging itemized ticket moves for analytics

### DISCUSSION THREAD

**Discussion by jstanden** [cerb/cerb-release#4] [Guides and Tutorials]
You can use [custom records](https://cerb.ai/docs/records/custom-records/), [bot behaviors](https://cerb.ai/docs/bots/), and [data queries](https://cerb.ai/docs/data-queries/) to track your own reporting metrics.

In this example, we'll create a custom record to log the movement of tickets between groups and buckets. This will also keep track of the date and the worker. A 'record changed' bot behavior will create new log records. We can then use data queries for reporting (e.g. _count of tickets moved from a Dispatch group by month_, _tickets moved by a specific worker_, _average tickets moved over time_, _move history of a specific ticket_).

## Import the example package

This package will automatically create the custom record, custom fields, and bot behavior.

1. Navigate to **Setup -> Packages -> Import**

2. Copy/paste:
~~~
{
  "package": {
    "name": "Move Log",
    "revision": 1,
    "requires": {
      "cerb_version": "9.6.0",
      "plugins": [

      ]
    },
    "configure": {
      "placeholders": [

      ],
      "prompts": [

      ]
    }
  },
  "bots": [
    {
      "uid": "bot_log",
      "name": "Log Bot",
      "owner": {
        "context": "cerberusweb.contexts.app",
        "id": 0
      },
      "is_disabled": false,
      "params": {
        "config": [

        ],
        "events": {
          "mode": "all",
          "items": [

          ]
        },
        "actions": {
          "mode": "all",
          "items": [

          ]
        }
      },
      "behaviors": [
        {
          "uid": "behavior_ticket_changed",
          "title": "Record ticket moves between groups",
          "is_disabled": false,
          "is_private": false,
          "priority": 50,
          "event": {
            "key": "event.record.changed",
            "label": "Record changed",
            "params": {
              "context": "cerberusweb.contexts.ticket"
            }
          },
          "nodes": [
            {
              "type": "switch",
              "title": "Is a ticket moving between groups?",
              "status": "live",
              "nodes": [
                {
                  "type": "outcome",
                  "title": "Yes",
                  "status": "live",
                  "params": {
                    "groups": [
                      {
                        "any": 0,
                        "conditions": [
                          {
                            "condition": "_custom_script",
                            "tpl": "{{old_group_id != group_id ? 'true'}}",
                            "oper": "is",
                            "value": "true"
                          }
                        ]
                      }
                    ]
                  },
                  "nodes": [
                    {
                      "type": "action",
                      "title": "Create a 'Move Log' record",
                      "status": "live",
                      "params": {
                        "actions": [
                          {
                            "action": "core.bot.action.record.create",
                            "context": "move_log",
                            "changeset_json": "{% set json = {\r\n  name: '%s moved `%s` to %s: %s'|format(actor__label, ticket__label, group__label, ticket_bucket__label),\r\n  owner__context: 'app',\r\n  owner_id: 0,\r\n  custom_{{{uid.field_from_group}}}: old_group_id,\r\n  custom_{{{uid.field_from_bucket}}}: old_ticket_bucket_id,\r\n  custom_{{{uid.field_to_group}}}: group_id,\r\n  custom_{{{uid.field_to_bucket}}}: ticket_bucket_id,\r\n  custom_{{{uid.field_ticket}}}: ticket_id,\r\n  custom_{{{uid.field_worker}}}: actor__context == 'cerberusweb.contexts.worker' ? actor_id : 0\r\n} %}\r\n{{json|json_encode|json_pretty}}",
                            "run_in_simulator": "0",
                            "object_placeholder": "_record"
                          }
                        ]
                      }
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ],
  "records": [
    {
      "uid": "record_type_move_log",
      "_context": "custom_record",
      "name": "Move Log",
      "name_plural": "Move Logs",
      "uri": "move_log",
      "params": {
        "owners": {
          "contexts": [
            "cerberusweb.contexts.app"
          ]
        }
      }
    },
    {
      "uid": "field_from_group",
      "_context": "custom_field",
      "name": "From Group",
      "context": "contexts.custom_record.{{{uid.record_type_move_log}}}",
      "type": "L",
      "pos": 1,
      "params": {
        "context": "cerberusweb.contexts.group"
      }
    },
    {
      "uid": "field_from_bucket",
      "_context": "custom_field",
      "name": "From Bucket",
      "context": "contexts.custom_record.{{{uid.record_type_move_log}}}",
      "type": "L",
      "pos": 2,
      "params": {
        "context": "cerberusweb.contexts.bucket"
      }
    },
    {
      "uid": "field_to_group",
      "_context": "custom_field",
      "name": "To Group",
      "context": "contexts.custom_record.{{{uid.record_type_move_log}}}",
      "type": "L",
      "pos": 3,
      "params": {
        "context": "cerberusweb.contexts.group"
      }
    },
    {
      "uid": "field_to_bucket",
      "_context": "custom_field",
      "name": "To Bucket",
      "context": "contexts.custom_record.{{{uid.record_type_move_log}}}",
      "type": "L",
      "pos": 4,
      "params": {
        "context": "cerberusweb.contexts.bucket"
      }
    },
    {
      "uid": "field_ticket",
      "_context": "custom_field",
      "name": "Ticket",
      "context": "contexts.custom_record.{{{uid.record_type_move_log}}}",
      "type": "L",
      "pos": 5,
      "params": {
        "context": "cerberusweb.contexts.ticket"
      }
    },
    {
      "uid": "field_worker",
      "_context": "custom_field",
      "name": "Worker",
      "context": "contexts.custom_record.{{{uid.record_type_move_log}}}",
      "type": "L",
      "pos": 6,
      "params": {
        "context": "cerberusweb.contexts.worker"
      }
    }
  ]
}
~~~

3. Click **Import**

You'll have a new record type:

<img width="715" alt="image" src="https://user-images.githubusercontent.com/63288/105779552-74065c00-5f23-11eb-87a8-47c8f46da41b.png">

And a bot behavior:

<img width="681" alt="image" src="https://user-images.githubusercontent.com/63288/105779487-5638f700-5f23-11eb-8f6c-6981869dd249.png">

## Test the behavior

1. Navigate to **Setup -> Mail -> Incoming -> Import**

2. Copy/paste the following message:
~~~
From: customer@cerb.example
To: support@cerb.example
Subject: Example ticket for testing moves

Ignore this ticket. It is being used to test new analytics.
~~~

3. Click the **Import** button.

4. Click on the new ticket link in the yellow alert.

5. From the ticket's profile page, edit and move it to a new group.

6. Check **Search -> show all -> Move Logs** to verify new log records are created.
  <img width="888" alt="image" src="https://user-images.githubusercontent.com/63288/105778904-2b01d800-5f22-11eb-869a-316b6b116223.png">

## Run reports with the new move log metrics

1. Navigate to **Setup -> Developers -> Data Query Tester**

2. Run this data query to subtotal ticket moves from a specific group over time:
  ~~~
  type:worklist.subtotals
  of:move_log
  by:[created@month,toGroup.id]
  query:(
    fromGroup:(name:Dispatch)
  )
  format:timeseries
  ~~~

3. Click the **Run** button.
  <img width="452" alt="image" src="https://user-images.githubusercontent.com/63288/105779940-42da5b80-5f24-11eb-92ca-4970cf69955a.png">

You can use data queries like this in chart widgets on [dashboards](https://cerb.ai/docs/dashboards/).

**Comment by mryanb** [comment]
Why might you choose to use a custom record over an activity log entry?

**Reply by jstanden** [reply]
Hey @mryanb!

With a custom record you can use key expansion on the record link fields. You can't access arbitrary properties from an activity log entry in search/data queries, etc.

With a custom record storing the move logs, you can easily create a report of tickets that moved *from* or *into* a specific group/bucket.

In this case, the client has a Triage group with human dispatchers that move all new tickets into other groups/buckets. They want metrics for destination group/buckets over time. They also want to keep track of dispatcher performance metrics.

At times, tickets are moved into the wrong group, which creates friction. This custom record would allow for feedback by group recipients whether a ticket was routed properly or not. Reports can be built to determine which dispatchers make the most defects, to which groups, during which shifts, etc.

Eventually, that feedback creates training data. A machine learning model could learn to predict what parameters of a new ticket message lead to ideal routing (e.g. subject key words, statistically significant phrases in the body text).

The custom record would then serve as a record of the ML prediction. Those could also be trained as good/bad by workers to refine the model.

The model could then assist human dispatchers in routing tickets; and when accurate enough, the model could do the routing itself.

The activity log is a good audit log of how a ticket came to be in its current group/bucket, but the log itself isn't actionable like the above.
